### PR TITLE
notifyParentDataSetChanged

### PR DESCRIPTION
**Pull request by paul-turner**
Implements the ability to call something similar to notifyDataSetChanged, with the added benefit of preserving expansion state. This PR is tracked by #89.

### CODE REVIEW

File: expandablerecyclerview/src/main/java/com/bignerdranch/expandablerecyclerview/ExpandableRecyclerAdapter.java [hunk 1]
```diff
@@ -66,6 +66,8 @@
     @NonNull
     private List<RecyclerView> mAttachedRecyclerViewPool;
 
+    private HashMap<P, Boolean> mExpansionStateMap;
```
Comment (dgreenhalgh) [inline expandablerecyclerview/src/main/java/com/bignerdranch/expandablerecyclerview/ExpandableRecyclerAdapter.java]: Prefer interface `Map` here

File: expandablerecyclerview/src/main/java/com/bignerdranch/expandablerecyclerview/ExpandableRecyclerAdapter.java [hunk 2]
```diff
@@ -337,6 +340,39 @@ public boolean isParentViewType(int viewType) {
...
+     *
+     * @see #notifyParentInserted(int)
+     * @see #notifyParentRemoved(int)
+     * @see #notifyParentChanged(int)
+     * @see #notifyParentRangeInserted(int, int)
+     * @see #notifyChildInserted(int, int)
+     * @see #notifyChildRemoved(int, int)
+     * @see #notifyChildChanged(int, int)
+     *
+     * @param preserveExpansionState If true, the adapter will preserve your parent's last expanded
```
Comment (dgreenhalgh) [inline expandablerecyclerview/src/main/java/com/bignerdranch/expandablerecyclerview/ExpandableRecyclerAdapter.java]: /s/preserve/attempt to preserve/

File: expandablerecyclerview/src/main/java/com/bignerdranch/expandablerecyclerview/ExpandableRecyclerAdapter.java [hunk 3]
```diff
@@ -760,6 +798,43 @@ int getChildPosition(int flatPosition) {
...
```
Comment (dgreenhalgh) [inline expandablerecyclerview/src/main/java/com/bignerdranch/expandablerecyclerview/ExpandableRecyclerAdapter.java]: /s/preserve/attempt to preserve/

File: expandablerecyclerview/src/main/java/com/bignerdranch/expandablerecyclerview/ExpandableRecyclerAdapter.java [hunk 4]
```diff
@@ -1238,32 +1313,64 @@ public void notifyChildMoved(int parentPosition, int fromChildPosition, int toCh
...
+    /**
+     * Generates a full list of all parents and their children, in order. Uses Hashmap to preserve
+     * last expanded state.
+     *
+     * @param parentList A list of the parents from
+     *                   the {@link ExpandableRecyclerAdapter}
+     * @param savedLastExpansionState A map of the last expanded state for a given parent key.
+     * @return A list of all parents and their children, expanded accordingly
+     */
+    private List<ExpandableWrapper<P, C>> generateFlattenedParentChildList(List<P> parentList, HashMap<P, Boolean> savedLastExpansionState) {
```
Comment (dgreenhalgh) [inline expandablerecyclerview/src/main/java/com/bignerdranch/expandablerecyclerview/ExpandableRecyclerAdapter.java]: Prefer interface `Map` here

**Comment by paul-turner** [comment]
👍  Made changes

**Comment by paul-turner** [comment]
fixes #239

**Comment by paul-turner** [comment]
fixes #185

**Pull request merged**